Unified validation check for dependency fields

dbochkov-flexcompute opened the pull request:
After playing a little bit with it, managed to implement what we wanted as a convenient decorator. The usage as follows: instead of doing something like 
```
@pydantic.validator("monitors", always=True)
def _warn_monitor_simulation_frequency_range(cls, val, values):
    """Warn if any DFT monitors have frequencies outside of the simulation frequency range."""

    if "sources" not in values:
        raise ValidationError(
            "could not validate `_warn_monitor_simulation_frequency_range` "
            "as `sources` failed validation"
        )

    ...
```
we can just do 
```
@pydantic.validator("monitors", always=True)
@check_previous_fields_validation(["sources"])
def _warn_monitor_simulation_frequency_range(cls, val, values):
    """Warn if any DFT monitors have frequencies outside of the simulation frequency range."""
    ...
```
this check whether dependency fields have passed validation, and, if not, it skips (not errors) the current validator and prints something like
```
WARNING: Could not execute validator                               
'_warn_monitor_simulation_frequency_range' because field 'sources' 
failed validation.
```

momchil-flex commented:
Why would we only warn instead of error? It seems to me that if previous fields failed validation we should error with the validation error of that field.

momchil-flex commented:
Also I think this PR should go to develop so that we patch this asap.

tylerflex commented on tidy3d/components/base.py:
```diff
@@ -86,6 +86,28 @@ def _get_valid_extension(fname: str) -> str:
+        def _validator(cls, val, values):
+            """New validator function."""
+            for field in required_fields:
+                if field not in values:
```
just to confirm, if `field` fails validation, it's just not put in `values`? (as opposed to put into values with `None`?)

dbochkov-flexcompute replied:
yeah, just double checked with pydantic documentation https://docs.pydantic.dev/1.10/usage/validators/:
> If validation fails on another field (or that field is missing) it will not be included in values, hence if 'password1' in values and ... in this example.

tylerflex commented on tidy3d/components/base.py:
```diff
@@ -86,6 +86,28 @@ def _get_valid_extension(fname: str) -> str:
     )
 
 
+def check_previous_fields_validation(required_fields):
```
can you type annotate `required_fields : List[str]`?

tylerflex commented on tidy3d/components/simulation.py:
there are probably a lot of cases in `Simulation` that could use this. For [example](https://github.com/flexcompute/tidy3d/blob/c0adac91959671fb07c9b41fb798e5e4a8de8272/tidy3d/components/simulation.py#L702-L705)

tylerflex reviewed (changes requested):
Thanks a lot @dbochkov-flexcompute, I think this will be great to have. A couple minor comments and also there are probably many instances here and there where we can implement it already, especially in `Simulation`. For example these [lines](https://github.com/flexcompute/tidy3d/blob/c0adac91959671fb07c9b41fb798e5e4a8de8272/tidy3d/components/simulation.py#L702-L705).

I wonder if, more generally, we can assume that if any `values.get(key)` or `values[key]` is present in a field validator, that we should probably `@check_previous_fields_validation` for that `key`.

Another pretty minor comment, perhaps the name of the decorator could be more specific, like
```
skip_if_validation_failed(fields: List[str])
```
because `check` is maybe a bit ambiguous as to what the decorator is doing. (Just a minor suggestion).

dbochkov-flexcompute commented:
> Why would we only warn instead of error? It seems to me that if previous fields failed validation we should error with the validation error of that field.

Just to be clear, the failed field still throws its own error. But I'm ok either way, the difference is very subtle. If we error, then pydantic will collect this additional error and overall we get the following output:
```
09:47:51 CST ERROR: Automatic grid generation requires the input of 'wavelength'
             or sources.
             ERROR: Could not execute validator 'plane_in_sim_bounds' because   
             field 'simulation' failed validation.
Traceback (most recent call last):
...
pydantic.v1.error_wrappers.ValidationError: 2 validation errors for ModeSolver
simulation -> grid_spec
  Automatic grid generation requires the input of 'wavelength' or sources. (type=value_error.setup)
plane
  Could not execute validator 'plane_in_sim_bounds' because field 'simulation' failed validation. (type=value_error)
```
Note that pydantic error list contains two errors (`pydantic.v1.error_wrappers.ValidationError: 2 validation errors for ModeSolver`).

If we just pop a warning, we get
```
16:58:19 CST ERROR: Automatic grid generation requires the input of 'wavelength'
             or sources.                                                        
             WARNING: Could not execute validator 'plane_in_sim_bounds' because 
             field 'simulation' failed validation.                              
Traceback (most recent call last):
...
pydantic.v1.error_wrappers.ValidationError: 1 validation error for ModeSolver
simulation -> grid_spec
  Automatic grid generation requires the input of 'wavelength' or sources. (type=value_error.setup)
```
Note that in this case pydantic error list contains just one error (`pydantic.v1.error_wrappers.ValidationError: 1 validation error for ModeSolver`).

This doesn't really make much difference, but I think when we explicitly error a field ("A") that was not validated just because a previous field ("B") failed validation, this might make an impression that something is wrong with setting up field "A", while in reality it could be that just resolving issues with filed "B" is enough. It might be more relevant for GUI were failed fields are marked visually.

momchil-flex commented:
Oh ok that makes sense thanks!

dbochkov-flexcompute commented:
revised this PR, bigger changes:
- rebased to develop
- per Tyler suggestion to make it more specific renamed the decorator to `skip_if_fields_missing`
- added it wherever it seemed to make sense

momchil-flex reviewed (approved):
Looks great to me!

tylerflex commented:
thanks @dbochkov-flexcompute
